js/src/ScatterGL.ts

```diff
@@ -0,0 +1,981 @@
+        this.attributes_active_previous.trim(this.attributes_active_previous.length);
+
+        if(this.attributes_active.length != this.attributes_active_previous.length) {
+            this.attributes_active.ensure_array('size')
+            this.attributes_active_previous.ensure_array('size')
+        }
+        if(this.attributes_active.length > this.attributes_active_previous.length) { // grow..
+            this.attributes_active_previous.pad(this.attributes_active)
+            this.attributes_active_previous.array['size'].fill(0, previous_length); // this will make them smoothly fade in
+        } else if(this.attributes_active.length < this.attributes_active_previous.length) { // shrink..
+            this.attributes_active.pad(this.attributes_active_previous)
+            this.attributes_active.array['size'].fill(0, current_length); // this will make them smoothly fade out
+        }
+        this.attributes_active.add_attributes(this.instanced_geometry)
+        this.attributes_active_previous.add_attributes(this.instanced_geometry, '_previous')
+
+        if(!this.mesh) {
+            this.mesh = new THREE.Mesh(this.instanced_geometry, this.scatter_material);
+        }
+        else {
+            this.mesh.geometry.dispose()
+            this.mesh.geometry = this.instanced_geometry;
+        }
+        _.each(attributes_changed, (key: any) => {
+            const property = "animation_time_" + key
+            const done = () => {
+                // _.each(attributes_changed, (prop) => {
+                    delete this.previous_values[key] // may happen multiple times, that is ok
+                // })
+                _.each(finalizers, (finalizer: any) => {
+                    finalizer()
+                })
+            }
+            // uniforms of material_rgb has a reference to these same object
+            const set = (value) => {
+                this.scatter_material.uniforms[property]['value'] = value
+            }
+            this.scatter_material.uniforms[property]['value'] = 0
+            this.transition(set, done, this)
+        })
+        this.attributes_changed = {}
+
+        this.update_scene();
+    }
+
+    update_scene() {
+        this.parent.update_gl()
+    }
+
+    render_gl() {
+        this.set_ranges()
+        this._update_requested = false;
+        const fig = this.parent;
+        this.im.attr('x', 0)
+               .attr('y', 0)
+               .attr('width', fig.plotarea_width)
+               .attr('height', fig.plotarea_height)
+        this.camera.left = 0
+        const x_scale = this.scales.x ? this.scales.x : this.parent.scale_x;
+        const y_scale = this.scales.y ? this.scales.y : this.parent.scale_y;
+
+        const range_x = this.parent.padded_range("x", x_scale.model);
+        const range_y = this.parent.padded_range("y", y_scale.model);
+
+        this.scatter_material.uniforms['colormap'].value = create_colormap(this.scales.color)
```

Comment:
Is it reasonable to create the colormap texture for each frame?

Another (more optimized?) strategy would be to create it only when the colorscale value changes.

Reply:
fixed